### PR TITLE
Allow struct fields to be accessible

### LINKED ISSUE #5: Issues with structs as variables
Hello,

I'm trying to use this library as I prefer it over Knetic's library based on features described, however, I'm struggling to get this working with structs. Consider the following scenario.
```go
type Testing struct {
	Title string
}
obj := &Testing{Title: "hello"}
eval := goval.NewEvaluator()
newrs, err := eval.Evaluate(`obj.Title == "hello"`, map[string]interface{}{"obj": obj}, nil)
# err == "syntax error: cannot access fields on type <unknown type>"
fmt.Printf(newrs.(string))
```

Where this scenario works fine:
```go
type Testing struct {
	Title string
}
//obj := &Testing{Title: "hello"}
eval := goval.NewEvaluator()
newrs, err := eval.Evaluate(`obj == "hello"`, map[string]interface{}{"obj": "hello"}, nil)
fmt.Printf(newrs.(string))
```

Any ideas how I can resolve this?

### PR DESCRIPTION
**Pull request by hoshsadiq**
This fixes #5

### CODE REVIEW

File: internal/parserUtils.go [hunk 1]
```diff
@@ -428,6 +428,27 @@ func accessField(s interface{}, field interface{}) interface{} {
...
+	for v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+	if v.Kind() == reflect.Struct {
+		key, ok := field.(string)
+		if !ok {
+			panic(fmt.Errorf("syntax error: object key must be string, but was %s", typeOf(field)))
+		}
+
+		if key[0] >= 'a' && key[0] <= 'z' {
```
Comment (maja42) [inline internal/parserUtils.go]: To ignore unexported fields, you should use 
```
	if !field.CanSet() {
		panic(fmt.Errorf("var error: object member %q is unexported", field))
	}
```

File: internal/parserUtils.go [hunk 2]
```diff
@@ -428,6 +428,27 @@ func accessField(s interface{}, field interface{}) interface{} {
...
 	}
 
+	v := reflect.ValueOf(s)
+	for v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+	if v.Kind() == reflect.Struct {
+		key, ok := field.(string)
+		if !ok {
+			panic(fmt.Errorf("syntax error: object key must be string, but was %s", typeOf(field)))
```
Comment (maja42) [inline internal/parserUtils.go]: The file already contains the function `asObjectKey` that can do that.

**Comment by hoshsadiq** [comment]
@maja42 think it's ready! let me know if you'd like any other changes

**Comment by hoshsadiq** [comment]
@maja42 any updates?

**Comment by maja42** [comment]
Hi, from a first glance it looks great. I'll merge it as soon as I possibly

**Comment by maja42** [comment]
The only thing missing right now are test-cases on how the individual operators are reacting to the new type.
The object-concatenation currently does not support structs (the add-function will need to be updated to either convert structs to maps or (better) create an error).

Also the "typeOf()" field will need to return "object" for structs - right now it returns <unknown type>.

In parser_test, the "getTestVars" function should also return a struct-member with content similar to the obj-member. And all unit-tests calling "getTestVars" should get an additional test entry for the struct type.

File: evaluator.go [hunk 3]
```diff
@@ -1,7 +1,7 @@
 package goval
 
 import (
-	"github.com/maja42/goval/internal"
+	"./internal"
```
Comment (hoshsadiq) [inline evaluator.go]: @maja42 did you mean to do this?

File: internal/parser_test.go [hunk 4]
```diff
@@ -1076,16 +1077,16 @@ func Test_VariableAccess_ArraySyntax_DoesNotExist(t *testing.T) {
...
-	assertEvalError(t, vars, "syntax error: object key must be string, but was bool", `obj[true]`)
-	assertEvalError(t, vars, "syntax error: object key must be string, but was number", `obj[0]`)
-	assertEvalError(t, vars, "syntax error: object key must be string, but was array", `obj[arr]`)
-	assertEvalError(t, vars, "syntax error: object key must be string, but was object", `obj[obj]`)
+	assertEvalError(t, vars, "type error: object key must be string, but was bool", `obj[true]`)
+	assertEvalError(t, vars, "type error: object key must be string, but was number", `obj[0]`)
+	assertEvalError(t, vars, "type error: object key must be string, but was array", `obj[arr]`)
+	assertEvalError(t, vars, "type error: object key must be string, but was object", `obj[obj]`)
 
-	assertEvalError(t, vars, "syntax error: array index must be number, but was bool", `arr[true]`)
```
Comment (hoshsadiq) [inline internal/parser_test.go]: In my opinion this is more of a syntax error than a type error.

**Comment by hoshsadiq** [comment]
Added those test cases

**Comment by hoshsadiq** [comment]
By the way, I replaced the string `"<unknown type>"` with `kind.String()`. This should make debugging easier. Let me know if you do not agree and I'll revert it.

File: evaluator.go (hunk 3 above)
Comment (hoshsadiq) [reply evaluator.go]: Yep, this breaks my code when I try to import it goval

Comment (maja42) [reply evaluator.go]: Oh, no that was an accident. I committed that on accident.

File: internal/parser_test.go (hunk 4 above)
Comment (maja42) [reply internal/parser_test.go]: The object key can also come from another variable:

`obj[var]`

This is syntactically correct. If `var` contains a number/float, the error is triggered. If it contains a string, everything is okay.
That's why I changed it to a type error.

File: evaluator.go (hunk 3 above)
Comment (hoshsadiq) [reply evaluator.go]: np, just pushed a fix!

**Comment by hoshsadiq** [comment]
@maja42 any update on this?

**Comment by maja42** [comment]
I tried to implement support for structs in this library, but eventually stopped working on it.
While it is easily possible to add basic support for structs, it creates a lot of loose ends and edge cases.

It would need to deal with pointers, interfaces, unexported members, member functions, nil-values, empty structs and so on. It would also need to integrate with already existing types like objects (maps) and arrays interchangably. If there are any edge-cases that aren't handled intuitively (and there will be a lot), the behaviour would need to be documented.

The thing I like most about this library compared to others is that the grammar is well-defined for all cases, and the existing types play together neatly. The whole syntax is easy to understand and won't cause suprises in the long run.

If you want to have struct-support, I'd recomment forking this package and taking a different approach. Currently, the expression is evaluated while being parsed by the lexer. To deal with the additional complexity, a better approach is to create an intermediate AST from the input and evaluate it afterwards.
This would reduce complexity in the yacc file and the actual logic can be written in go.
It would also allow more "sophisticated" syntax like a ternery operator or boolean operators that don't evaluate every component once the outcome is already clear (`a() && b()` will always call both functions, even if `a` returned false).

**Pull request closed without merging**